Import a method from an python_print string

zdevito opened the pull request:
* Add hooks to get a callback whenever a valid graph is produced in the compiler or through tracing. These hooks can be used to pretty_print and then reparse every graph our tests produce to check that the serialization function works correctly. Currently this is guarded by an environment variable since there are a few remaining failures.
* Fix printing bugs: True and False rather than 1 and 0, print 0. for floating point zero
* Change behavior of NoneType. It is now no longer a subtype of Optional but instead implicitly converts to it, returning a prim::Node with an Option[T] type for some specific T. This allows functions like `_unwrap_optional` to correctly match against a None while still deriving the right type. 
* Fix a bug where empty blocks did not correctly emit "pass" in printer.
* Fix a bug where prim::Undefine sometimes cannot be printed as None because it is being used in a schema-less op. This should be fixable once Optional[T] always uses the same None object.
* Other minor printing bugs

wanchaol reviewed (commented):
See a nccl submodule update accidentally checked in, also might be better if we can add more test cases to cover things like NoneType changes.

zdevito commented on torch/csrc/jit/import_method.cpp:
```diff
@@ -0,0 +1,113 @@
+
+// this is a much simpler accessor that only handles modules, parameters, and
+// and methods
+struct ModuleAccessorValue : public script::SugaredValue {
```
ModuleValue depends on libpython.so, but this file is used in libtorch.so where no python is available.

zdevito commented on test/test_jit.py:
```diff
@@ -2079,6 +2106,16 @@ def foo(a):
+
+        r = foo.graph.pretty_print()
+        mod = torch.jit.ScriptModule()
+        torch._C._jit_import_method(mod, "op_version_set = 0\n{}".format(r), [])
```
it is not part of the python_print functionality. It will be part of the module-level printing, so I just add it here to let import work.

zdevito commented on torch/csrc/jit/autodiff.cpp:
```diff
@@ -177,15 +178,20 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
     } else if (node->matches("aten::clamp(Tensor self, Scalar? min, Scalar? max) -> Tensor")) {
       // handle the case that min/max is None
       Value* min = inputs.at(1);
+      bool min_must_be_none = min->node()->kind() == prim::None;
```
In person: this use is currently broken, and we've added it to the task list. To cause the problem the example needs to look like:
```
@torch.jit.script
def test(input, min, max):
    # type: (Tensor, Optional[float], Optional[float]) -> Tensor
    t = input.clamp(min, max)
    return t + t + t
test(torch.randn(3,4), 0.0, None)
```

zdevito commented on torch/csrc/jit/passes/python_print.cpp:
```diff
@@ -504,19 +519,75 @@ class PythonPrintPass {
+          stmt << "None";
+          break;
+        }
+        // XXX - when None has an Optional[T] type, we must ensure that type
```
discussed in person: because of the difficulty of recovering the T in an `unpack_optional(None)`,  we need to make `prim::None` have the ability to have type `optional[T]` for some `T`.

wanchaol commented on torch/csrc/jit/import_method.cpp:
```diff
@@ -0,0 +1,113 @@
+  std::shared_ptr<script::Module> module;
+};
+
+// this is a much simpler accessor that only handles modules, parameters, and
```
Can we have a more proper comment on the ConstantTableValue, it is not quite intuitive what is it for and how it's being used.

wanchaol commented on torch/csrc/jit/passes/python_print.h:
```diff
@@ -1,13 +1,18 @@
 namespace torch { namespace jit {
 struct Graph;
-TORCH_API std::ostream& PythonPrint(std::ostream& out, const Graph& graph);
+TORCH_API std::vector<at::Tensor> PythonPrint(std::ostream& out, const Graph& graph, bool enforce_importable=false);
```
This looks a little bit weird on the PythonPrint interface, As we are doing import/export, the constant table is not getting serialized/deserialized, so basically we might lose some information on importing if no constant table is preserved (emitModule hook are storing them in memory, not in disk).

zdevito replied:
import_method just handles importing the code. There will be another piece of code that handles encoding the constant table and restoring it. These hooks test the import mechanism for code separately from the end-to-end module tests which do not exist yet.

wanchaol reviewed (approved):
LGTM after in person discussion. NoneType behavior change is our current compromise on recovering the proper optional type information.

zdevito committed:
Import a method from an python_print string (#13959)

Summary:
* Add hooks to get a callback whenever a valid graph is produced in the compiler or through tracing. These hooks can be used to pretty_print and then reparse every graph our tests produce to check that the serialization function works correctly. Currently this is guarded by an environment variable since there are a few remaining failures.
* Fix printing bugs: True and False rather than 1 and 0, print 0. for floating point zero
* Change behavior of NoneType. It is now no longer a subtype of Optional but instead implicitly converts to it, returning a prim::Node with an Option[T] type for some specific T. This allows functions like `_unwrap_optional` to correctly match against a None while still deriving the right type.
* Fix a bug where empty blocks did not correctly emit "pass" in printer.
* Fix a bug where prim::Undefine sometimes cannot be printed as None because it is being used in a schema-less op. This should be fixable once Optional[T] always uses the same None object.
* Other minor printing bugs
Pull Request resolved: https://github.com/pytorch/pytorch/pull/13959

Reviewed By: wanchaol

Differential Revision: D13073519

Pulled By: wanchaol

fbshipit-source-id: d1eacd8ecf60e06bf356b4d5814b6c5ba9eb06df
